core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -244,13 +245,34 @@ public AllocateUnassignedDecision makeAllocationDecision(final ShardRouting unas
     /**
      * Builds a map of nodes to the corresponding allocation decisions for those nodes.
      */
-    private static List<NodeAllocationResult> buildNodeDecisions(NodesToAllocate nodesToAllocate, Set<String> inSyncAllocationIds) {
-        return Stream.of(nodesToAllocate.yesNodeShards, nodesToAllocate.throttleNodeShards, nodesToAllocate.noNodeShards)
-                   .flatMap(Collection::stream)
-                   .map(dnode -> new NodeAllocationResult(dnode.nodeShardState.getNode(),
-                                                             shardStoreInfo(dnode.nodeShardState, inSyncAllocationIds),
-                                                             dnode.decision))
-                   .collect(Collectors.toList());
+    private static List<NodeAllocationResult> buildNodeDecisions(NodesToAllocate nodesToAllocate,
+                                                                 FetchResult<NodeGatewayStartedShards> fetchedShardData,
+                                                                 Set<String> inSyncAllocationIds) {
+        List<NodeAllocationResult> nodeResults = new ArrayList<>();
+        Collection<NodeGatewayStartedShards> ineligibleShards;
+        if (nodesToAllocate != null) {
+            nodeResults.addAll(Stream.of(nodesToAllocate.yesNodeShards, nodesToAllocate.throttleNodeShards, nodesToAllocate.noNodeShards)
+                                .flatMap(Collection::stream)
+                                .map(dnode -> new NodeAllocationResult(dnode.nodeShardState.getNode(),
+                                                                       shardStoreInfo(dnode.nodeShardState, inSyncAllocationIds),
+                                                                       dnode.decision))
+                                .collect(Collectors.toList()));
+            ineligibleShards = fetchedShardData.getData().values().stream().filter(shardData ->
+                nodesToAllocate.yesNodeShards.contains(shardData) == false
```

Comment:
this is extremely inefficient (using contains on a list). Maybe it's sufficient to collect the `DiscoveryNode`s of yes/no/throttle into a set and do a lookup on that?